From f39e4949ffe3392289d7b26d5bb02acf75a7fea5 Mon Sep 17 00:00:00 2001
From: Zhihai Xu <zxu@apache.org>
Date: Thu, 1 Oct 2015 11:56:49 -0700
Subject: [PATCH 1687/2848] HADOOP-8437. getLocalPathForWrite should throw
 IOException for invalid paths. Contributed by
 Brahma Reddy Battula

(cherry picked from commit fd026f535cc09e99a7d4d5d2a8c13eabe8865315)

Change-Id: I8ba38c44514c32be0d631b386719ab91ec8323e3
---
 .../org/apache/hadoop/fs/LocalDirAllocator.java    |    6 ++++--
 .../apache/hadoop/fs/TestLocalDirAllocator.java    |   17 +++++++++++++++++
 2 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/LocalDirAllocator.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/LocalDirAllocator.java
index 88b4d4e..70e671f 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/LocalDirAllocator.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/LocalDirAllocator.java
@@ -304,8 +304,10 @@ private synchronized void confChanged(Configuration conf)
         dirDF = dfList.toArray(new DF[dirs.size()]);
         savedLocalDirs = newLocalDirs;
         
-        // randomize the first disk picked in the round-robin selection 
-        dirNumLastAccessed = dirIndexRandomizer.nextInt(dirs.size());
+        if (dirs.size() > 0) {
+          // randomize the first disk picked in the round-robin selection
+          dirNumLastAccessed = dirIndexRandomizer.nextInt(dirs.size());
+        }
       }
     }
 
diff --git a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/TestLocalDirAllocator.java b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/TestLocalDirAllocator.java
index 624fa14..8690205 100644
--- a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/TestLocalDirAllocator.java
+++ b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/TestLocalDirAllocator.java
@@ -434,4 +434,21 @@ public void testRemoveContext() throws IOException {
     }
   }
 
+  /**
+   * Test to check the LocalDirAllocation for the invalid path HADOOP-8437
+   *
+   * @throws Exception
+   */
+  @Test(timeout = 30000)
+  public void testGetLocalPathForWriteForInvalidPaths() throws Exception {
+    conf.set(CONTEXT, " ");
+    try {
+      dirAllocator.getLocalPathForWrite("/test", conf);
+      fail("not throwing the exception");
+    } catch (IOException e) {
+      assertEquals("Incorrect exception message",
+          "No space available in any of the local directories.", e.getMessage());
+    }
+  }
+
 }
-- 
1.7.9.5

